File: data-structure-and-algorithm/oc/strobogrammatic-number-II.m
// 247. Strobogrammatic Number II
// Difficulty : Medium

// A strobogrammatic number is a number that looks the same when rotated 180 degrees (looked at upside down).

// Find all strobogrammatic numbers that are of length = n.

// For example,
// Given n = 2, return ["11","69","88","96"].

// Hint:
// Try to use recursion and notice that it should recurse with n - 2 instead of n - 1

// Time:  O(n^2 * 5^(n/2))
// Space: O(n)

#import <Foundation/Foundation.h>

NSArray* findStrobogrammaticHelper(const int n, int k, NSDictionary* lookup) {
  if (k == 0) {
    return @[@""];
  } else if (k == 1) {
    return @[@"0", @"1", @"8"];
  }
  NSMutableArray* result = @[].mutableCopy;
  for (id num in findStrobogrammaticHelper(n, k - 2, lookup)) {
    for (id key in lookup) {
      if (n != k || ![key isEqual:@"0"]) {
        [result addObject:[NSString stringWithFormat:@"%@%@%@", key, num, lookup[key]]];
      }
    }
  }
  return result;
}

NSArray* findStrobogrammatic(int n) {
  NSDictionary* lookup = @{@"0":@"0", @"1":@"1", @"6":@"9", @"8":@"8", @"9":@"6"};
  return findStrobogrammaticHelper(n, n, lookup);
}
  

File: data-structure-and-algorithm/typescript/leetcode/src/strobogrammatic-number-ii.ts
/*
A strobogrammatic number is a number that looks the same when rotated 180 degrees (looked at upside down).

Find all strobogrammatic numbers that are of length = n.

Example:

Input:  n = 2
Output: ["11","69","88","96"]
*/

function findStrobogrammatic(size: number): string[] {
  const pairMap = new Map<string, string>([
    ['0', '0'],
    ['1', '1'],
    ['8', '8'],
    ['6', '9'],
    ['9', '6'],
  ]);
  return findStrobogrammaticDfs(size, size);

  function findStrobogrammaticDfs(size: number, count: number): string[] {
    if (count === 0) {
      return [''];
    }
    if (count === 1) {
      return ['0', '1', '8'];
    }

    const path: string[] = [];
    for (const mid of findStrobogrammaticDfs(size, count - 2)) {
      for (const [left, right] of pairMap) {
        if (count !== size || left !== '0') {
          path.push(left + mid + right);
        }
      }
    }
    return path;
  }
}
